[stdlib] Use addressor for Dictionary's subscript(_:default:)

hamishknight opened the pull request:
Given we now have the law of exclusivity, I believe we can use an addressor for `Dictionary`'s `subscript(_:default:)` by first inserting the default value if not already present for the given key, and handing back a pointer to the value in the buffer.

This allows us to prevent the copying of a dictionary's value on mutation through this subscript. For example, we can now mutate an array value in-place:

    var dict: [String: [String]] = ["foo": []]
    dict["foo", default: []].append("bar") // in-place appending
    dict["foo", default: []].append("baz") // in-place appending

The only cons that I can see to this change is that:

1. Using the subscript for plain assignment (`dict["foo", default: []] = ...`) is now less efficient, as it'll attempt to insert the default value before immediately overwriting it with the new value (and needlessly evaluate the autoclosure'd default value if inserted). However, frankly using assignment with this subscript is pointless, as the default value has no effect.

2. Code that runs without exclusivity checks will be able to see the default value already inserted while in the window of mutation. However I don't *believe* we guarantee any specified behaviour for breaking the law of exclusivity (is that right?).

I appreciate any and all feedback on this change; specifically whether or not I've missed any reason why we cannot/shouldn't do this.

hamishknight committed:
[stdlib] Use addressor for Dictionary's subscript(_:default:)

hamishknight commented on test/stdlib/Inputs/DictionaryKeyValueTypes.swift:
```diff
@@ -25,6 +25,31 @@ func equalsUnordered<T : Comparable>(_ lhs: [T], _ rhs: [T]) -> Bool {
   return lhs.sorted().elementsEqual(rhs.sorted())
 }
 
+// A COW wrapper type that holds an Int.
```
Is this the right place to define this, and is there an already existing mock COW wrapper I could use here?

lorentey replied:
It's perfectly fine to add this here!

hamishknight commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please benchmark

natecook1000 commented:
@hamishknight This is super interesting! It looks like this might also solve an issue where using a class instance as the default doesn't actually assign:

```swift
var d: [String: UIView] = [:]
d["foo", default: UIView()].addSubview(UIView())
print(d["foo"] as Any)
```

Also, re this bit, the CI has been updated so it can handle new benchmarks in this same PR:

> I need to open a PR that adds subscript(_:default:) benchmark(s) so we can see an improvement on this PR being merged.

hamishknight commented:
@natecook1000 Thanks! I don't believe it'll solve the problem you mention as the compiler will still just call the getter directly (as it's a reference type), therefore not inserting the default value. I believe that'd have to be a compiler-level change to make that work.

> Also, re this bit, the CI has been updated so it can handle new benchmarks in this same PR:

>> I need to open a PR that adds subscript(_:default:) benchmark(s) so we can see an improvement on this PR being merged.

Awesome – I'll get on with that then :)

hamishknight committed:
[stdlib] Added benchmarks for Dictionary's subscript(_:default:)

hamishknight commented:
@natecook1000 I've gone ahead and added the benchmarks, could you please kick off the CI bot? (I'm not sure whether or not I'm able to). I think this PR is ready for review now :)

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke benchmark

natecook1000 commented:
@swift-ci Please benchmark

natecook1000 commented:
@swift-ci Please benchmark

slavapestov commented:
Some nice performance gains here!

hamishknight committed:
[stdlib] Fix uniqueness check in nativePointerToValue(forKey:insertingDefault:)

The reasoning for this change is detailed in PR-13000.

hamishknight commented:
Just fixed the uniqueness check in `nativePointerToValue(forKey:insertingDefault:)` in accordance with https://github.com/apple/swift/pull/13000.

Any chance of getting this moving soon?

hamishknight committed:
[stdlib] Don't use autoclosure on pointerToValue(forKey:insertingDefault:) methods

Previously we used @autoclosure on the _VariantDictionaryBuffer methods and forwarded closures directly.
However, this relied on an unintentional type-checker inconsistency, tracked by SR-5719.

natecook1000 commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2048,7 +2048,7 @@ public struct Dictionary<Key : Hashable, Value> :
     mutableAddressWithNativeOwner {
       let (_, address) = _variantBuffer
-        .pointerToValue(forKey: key, insertingDefault: defaultValue)
+        .pointerToValue(forKey: key, insertingDefault: defaultValue())
```
Could we remove the `@autoclosure` attribute from the `_VariantDictionaryBuffer` methods instead? Without being able to see the signature for `pointerToValue` it really looks like `defaultValue` is getting called here.

hamishknight replied:
I didn't have any strong opinion on it at the time; but now you mention it, yup, I agree.

hamishknight commented:
Any news on this @moiseev?

moiseev commented:
Oops. Sorry did not pay attention.
@lorentey, you've been working with dictionaries recently. Mind taking a look?
/cc @airspeedswift

hamishknight committed:
[stdlib] Don't check for native when known native in ensureNativeBuffer()

airspeedswift commented:
The trouble is we _don't_ have the law of exclusivity, at least not for all code, it's still opt-in. So if a user has it turned off we need to be very careful about exactly what the situation is in that case. If it's just odd but safe behaviors in case of situations that area already considered invalid that might be ok, unsafe/undefined behavior would not. cc @atrick / @rjmccall

hamishknight commented:
@airspeedswift This should fall under "odd but safe behavior" – the only difference to a user running code without exclusivity checks should be that during the window of mutation of a dictionary through `subscript(_:default:)`, they'll potentially be able to see the default value already inserted (whereas before; they would only see the value inserted afterwards). I don't know whether anyone is actually relying on the previous behaviour though.

Do we actually guarantee specified behaviour for breaking the law of exclusivity though? From what I could tell, the stance towards that was it mustn't introduce undefined behaviour, but the user should expect their program to be difficult to reason about (is this correct?). IMO, this change fits well with that.

Although aren't exclusivity checks currently opt-out? `-enforce-exclusivity=checked` seems to be the default for `swiftc`.

rjmccall commented:
Because we do not yet have guaranteed exclusivity, you need to use an owning addressor in order to ensure that nobody destroys or structurally modifies the buffer during the subscript access.  But since you *are* using an owning addressor, I think this should be fine.

lorentey commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -5338,7 +5338,16 @@ internal enum _Variant${Self}Buffer<${TypeParametersDecl}> : _HashBuffer {
+      //
+      // SR-6437
+      let capacity = asNative.capacity
+      _ = ensureUniqueNativeBuffer(capacity)
```
Shouldn't we just call _nativePointerToValue(at:) here?

hamishknight replied:
Thanks @lorentey, I've gone ahead and made that change.

lorentey commented:
@swift-ci Please test

hamishknight committed:
[stdlib] Call nativePointerToValue(at:) from nativePointerToValue(forKey:insertingDefault:)

lorentey commented:
@swift-ci Please test
